File: paas-metasvr/src/main/java/com/zzstack/paas/underlying/metasvr/autodeploy/deployer/RedisClusterDeployer.java
package com.zzstack.paas.underlying.metasvr.autodeploy.deployer;

import com.zzstack.paas.underlying.metasvr.autodeploy.ServiceDeployer;
import com.zzstack.paas.underlying.metasvr.autodeploy.util.CollectdDeployUtils;
import com.zzstack.paas.underlying.metasvr.autodeploy.util.DeployUtils;
import com.zzstack.paas.underlying.metasvr.autodeploy.util.InstanceOperationEnum;
import com.zzstack.paas.underlying.metasvr.autodeploy.util.RedisDeployUtils;
import com.zzstack.paas.underlying.metasvr.bean.PaasInstance;
import com.zzstack.paas.underlying.metasvr.bean.PaasMetaCmpt;
import com.zzstack.paas.underlying.metasvr.bean.PaasRedisCluster;
import com.zzstack.paas.underlying.metasvr.bean.PaasService;
import com.zzstack.paas.underlying.metasvr.bean.PaasSsh;
import com.zzstack.paas.underlying.metasvr.bean.TopoResult;
import com.zzstack.paas.underlying.metasvr.consts.FixDefs;
import com.zzstack.paas.underlying.metasvr.global.DeployLog;
import com.zzstack.paas.underlying.metasvr.metadata.CmptMeta;
import com.zzstack.paas.underlying.metasvr.singleton.MetaSvrGlobalRes;
import com.zzstack.paas.underlying.metasvr.utils.StringUtils;
import com.zzstack.paas.underlying.utils.FixHeader;
import com.zzstack.paas.underlying.utils.bean.ResultBean;
import com.zzstack.paas.underlying.utils.consts.CONSTS;

import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

public class RedisClusterDeployer implements ServiceDeployer {
    
    @Override
    public boolean deployService(String servInstID, String deployFlag, String logKey, String magicKey, ResultBean result) {
        TopoResult topoResult = DeployUtils.LoadServTopo(servInstID, logKey, true, result);
        if (!topoResult.isOk()) {
            return false;
        }
        JsonObject servJson = topoResult.getServJson();
        String version = topoResult.getVersion();

        if (CONSTS.DEPLOY_FLAG_PSEUDO.equals(deployFlag)) {
            if (!RedisDeployUtils.deployFakeClusterService(servJson, result, servInstID, magicKey)) {
                DeployLog.pubFailLog(logKey, "redis deploy failed ......");
                return false;
            }

            String info = String.format("service inst_id:%s, deploy sucess ......", servInstID);
            DeployLog.pubSuccessLog(logKey, info);
            return true;
        }

        // JsonObject collectdJson = servJson.getJsonObject(FixHeader.HEADER_COLLECTD);
        JsonObject nodeContainer = servJson.getJsonObject(FixHeader.HEADER_REDIS_NODE_CONTAINER);
        JsonObject proxyContainer = servJson.getJsonObject(FixHeader.HEADER_REDIS_PROXY_CONTAINER);

        JsonArray redisNodeArr = nodeContainer.getJsonArray(FixHeader.HEADER_REDIS_NODE);
        JsonArray proxyArr = proxyContainer.getJsonArray(FixHeader.HEADER_REDIS_PROXY);

        StringBuilder node4cluster = new StringBuilder("");
        StringBuilder nodes4proxy = new StringBuilder("");
        if (!RedisDeployUtils.getClusterNodes(redisNodeArr, node4cluster, nodes4proxy, logKey, result)) return false;
        
        // 1. deploy redis nodes
        int redisNodeCnt = redisNodeArr.size();
        for (int idx = 0; idx < redisNodeCnt; ++idx) {
            boolean init = idx == (redisNodeCnt - 1);
            JsonObject redisJson = redisNodeArr.getJsonObject(idx);
            if (!RedisDeployUtils.deployRedisNode(redisJson, init, false, true, node4cluster.toString(), version,
                    logKey, magicKey, result))
                return false;
        }

        // 2. deploy proxy
        int redisProxyCnt = proxyArr.size();
        for (int idx = 0; idx < redisProxyCnt; ++idx) {
            JsonObject proxyJson = proxyArr.getJsonObject(idx);
            if (!RedisDeployUtils.deployProxyNode(proxyJson, nodes4proxy.toString(), logKey, magicKey, result))
                return false;
        }

        // 部署collectd服务
        if (servJson.containsKey(FixHeader.HEADER_COLLECTD)) {
            JsonObject collectd = servJson.getJsonObject(FixHeader.HEADER_COLLECTD);
            if (!collectd.isEmpty()) {
                if (!CollectdDeployUtils.deployCollectd(collectd, servInstID, logKey, magicKey, result)) {
                    DeployLog.pubFailLog(logKey, "collectd start failed ......");
                    return false;
                }
            }
        }

        // update deploy flag and local cache
        DeployUtils.postProc(servInstID, FixDefs.STR_TRUE, logKey, magicKey, result);
        return true;
    }


    @Override
    public boolean undeployService(String servInstID, boolean force, String logKey, String magicKey, ResultBean result) {
        TopoResult topoResult = DeployUtils.LoadServTopo(servInstID, logKey, false, result);
        if (!topoResult.isOk()) {
            return false;
        }
        JsonObject servJson = topoResult.getServJson();
        
        CmptMeta cmptMeta = MetaSvrGlobalRes.get().getCmptMeta();
        PaasService serv = cmptMeta.getService(servInstID);

        //卸载伪部署
        if (CONSTS.DEPLOY_FLAG_PSEUDO.equals(serv.getPseudoDeployFlag())) { //服务是伪部署的话
            if (!RedisDeployUtils.unDeployFakeClusterService(servJson, result, servInstID, magicKey)) {
                return false;
            }

            String info = String.format("service inst_id: %s, undeploy sucess ......", servInstID);
            DeployLog.pubSuccessLog(logKey, info);
            return true;
        }

        // JsonObject collectdJson = servJson.getJsonObject(FixHeader.HEADER_COLLECTD);
        JsonObject nodeContainer = servJson.getJsonObject(FixHeader.HEADER_REDIS_NODE_CONTAINER);
        JsonObject proxyContainer = servJson.getJsonObject(FixHeader.HEADER_REDIS_PROXY_CONTAINER);

        JsonArray redisNodeArr = nodeContainer.getJsonArray(FixHeader.HEADER_REDIS_NODE);
        JsonArray proxyArr = proxyContainer.getJsonArray(FixHeader.HEADER_REDIS_PROXY);
        
        // 1. undeploy proxy
        int proxySize = proxyArr.size();
        for (int idx = 0; idx < proxySize; ++idx) {
            JsonObject proxyJson = proxyArr.getJsonObject(idx);
            if (!RedisDeployUtils.undeployProxyNode(proxyJson, force, logKey, magicKey, result)) {
                if (!force) return false;
            }
        }
        
        // 2. undeploy redis nodes
        int redisNodeSize = redisNodeArr.size();
        for (int idx = 0; idx < redisNodeSize; ++idx) {
            JsonObject redisJson = redisNodeArr.getJsonObject(idx);
            
            if (!RedisDeployUtils.undeployRedisNode(redisJson, false, logKey, magicKey, result)) {
                if (!force) return false;
            }
        }
        // 卸载collectd服务
        if (servJson.containsKey(FixHeader.HEADER_COLLECTD)) {
            JsonObject collectd = servJson.getJsonObject(FixHeader.HEADER_COLLECTD);
            if (!collectd.isEmpty()) {
                if (!CollectdDeployUtils.undeployCollectd(collectd, logKey, magicKey, result)) {
                    DeployLog.pubFailLog(logKey, "collectd undeploy failed ......");
                    return false;
                }
            }
        }
        
        // update deploy flag and local cache
        DeployUtils.postProc(servInstID, FixDefs.STR_FALSE, logKey, magicKey, result);
        return true;
    }

    @Override
    public boolean deployInstance(String servInstID, String instID, String logKey, String magicKey, ResultBean result) {
        TopoResult topoResult = DeployUtils.LoadServTopo(servInstID, logKey, false, result);
        if (!topoResult.isOk()) {
            return false;
        }
        JsonObject servJson = topoResult.getServJson();
        String version = topoResult.getVersion();

        JsonObject nodeContainer = servJson.getJsonObject(FixHeader.HEADER_REDIS_NODE_CONTAINER);
        JsonObject proxyContainer = servJson.getJsonObject(FixHeader.HEADER_REDIS_PROXY_CONTAINER);

        JsonArray redisNodeArr = nodeContainer.getJsonArray(FixHeader.HEADER_REDIS_NODE);
        JsonArray proxyArr = proxyContainer.getJsonArray(FixHeader.HEADER_REDIS_PROXY);
        
        StringBuilder node4cluster = new StringBuilder();
        StringBuilder nodes4proxy = new StringBuilder();
        if (!RedisDeployUtils.getClusterNodes(redisNodeArr, node4cluster, nodes4proxy, logKey, result)) return false;
        
        CmptMeta cmptMeta = MetaSvrGlobalRes.get().getCmptMeta();
        PaasInstance inst = cmptMeta.getInstance(instID);
        PaasMetaCmpt instCmpt = cmptMeta.getCmptById(inst.getCmptId());
        boolean deployResult = false;

        switch (instCmpt.getCmptName()) {
        case FixDefs.CMPT_REDIS_PROXY:
            JsonObject proxyJson = DeployUtils.getSpecifiedItem(proxyArr, instID);
            deployResult = RedisDeployUtils.deployProxyNode(proxyJson, nodes4proxy.toString(), logKey, magicKey, result);
            break;
        case FixDefs.CMPT_REDIS_NODE:
            String joinIp = "", joinPort = "", sshUser = "", sshPasswd = "";
            int sshPort = 22;
            int nodeSize = redisNodeArr.size();
            for (int i = 0; i < nodeSize; ++i) {
                JsonObject redis_json = redisNodeArr.getJsonObject(i);
                String curr_id = redis_json.getString(FixHeader.HEADER_INST_ID);
                if (!curr_id.equals(instID)) {
                    if (joinPort.length() == 0) {
                        // get some one node in cluster in order to connect and fetch cluster info.
                        // fetch_ip_and_port(join_ip, join_port, ssh_user, ssh_passwd, ssh_port, redis_json, logKey, result);
                        String sshID = redis_json.getString(FixHeader.HEADER_SSH_ID);
                        PaasSsh ssh = DeployUtils.getSshById(sshID, logKey, result);
                        if (ssh != null) {
                            joinIp    = ssh.getServerIp();
                            sshUser   = ssh.getSshName();
                            sshPasswd = ssh.getSshPwd();
                            sshPort   = ssh.getSshPort();
                            joinPort  = redis_json.getString(FixHeader.HEADER_PORT);
                        }
                    }
                    continue;
                }
                
                if (!RedisDeployUtils.deployRedisNode(redis_json, false, true, true, "", version, logKey, magicKey,
                        result))
                    return false;

                // get deployed cluster nodes info
                ResultBean clusterInfo = new ResultBean();
                if (!RedisDeployUtils.getRedisClusterNode(joinIp, joinPort, sshUser, sshPasswd, sshPort, logKey, clusterInfo)) return false;
                PaasRedisCluster cluster = new PaasRedisCluster();
                cluster.parse(clusterInfo.getRetInfo());
                // get self node info.
                String selfSshID = redis_json.getString(FixHeader.HEADER_SSH_ID);
                PaasSsh selfSsh = DeployUtils.getSshById(selfSshID, logKey, result);
                if (selfSsh != null) {
                    joinIp    = selfSsh.getServerIp();
                    sshUser   = selfSsh.getSshName();
                    sshPasswd = selfSsh.getSshPwd();
                    sshPort   = selfSsh.getSshPort();
                    joinPort  = redis_json.getString(FixHeader.HEADER_PORT);
                }

                // first check whether REDIS_CLUSTER_REPLICAS > 0 and have mater node with no slave
                // if REDIS_CLUSTER_REPLICAS == 0 new node is joined with master role,
                // or else need to check cluster status to jutify which role.

                // get deployed cluster nodes info
                
                // join cluster and reshard
                if (FixDefs.REDIS_CLUSTER_REPLICAS > 0) {
                    String alone_master_id = cluster.getAloneMaster();
                    if (StringUtils.isNull(alone_master_id)) {
                        // join as master node
                        if (!RedisDeployUtils.joinAsMasterNode(joinIp, joinPort, sshUser, sshPasswd, sshPort, cluster, logKey, result)) return false;
                    } else {
                        // join as slave node
                        if (!RedisDeployUtils.joinAsSlaveNode(joinIp, joinPort, sshUser, sshPasswd, sshPort, cluster, logKey, result)) return false;
                    }
                }
            }
            break;
        case FixDefs.CMPT_COLLECTD:
            JsonObject collectd = servJson.getJsonObject(FixHeader.HEADER_COLLECTD);
            deployResult = CollectdDeployUtils.deployCollectd(collectd, servInstID, logKey, magicKey, result);
            break;
        default:
            break;
        }

        DeployUtils.postDeployLog(deployResult, servInstID, logKey, "deploy");
        return deployResult;
    }

    @Override
    public boolean undeployInstance(String servInstID, String instID, String logKey, String magicKey, ResultBean result) {
        TopoResult topoResult = DeployUtils.LoadServTopo(servInstID, logKey, false, result);
        if (!topoResult.isOk()) {
            return false;
        }
        JsonObject servJson = topoResult.getServJson();
        
        JsonObject nodeContainer = servJson.getJsonObject(FixHeader.HEADER_REDIS_NODE_CONTAINER);
        JsonObject proxyContainer = servJson.getJsonObject(FixHeader.HEADER_REDIS_PROXY_CONTAINER);

        JsonArray redisNodeArr = nodeContainer.getJsonArray(FixHeader.HEADER_REDIS_NODE);
        JsonArray proxyArr = proxyContainer.getJsonArray(FixHeader.HEADER_REDIS_PROXY);
        
        CmptMeta cmptMeta = MetaSvrGlobalRes.get().getCmptMeta();
        PaasInstance inst = cmptMeta.getInstance(instID);
        PaasMetaCmpt instCmpt = cmptMeta.getCmptById(inst.getCmptId());
        boolean undeployResult = false;

        switch (instCmpt.getCmptName()) {
        case FixDefs.CMPT_REDIS_PROXY:
            JsonObject proxyJson = DeployUtils.getSpecifiedItem(proxyArr, instID);
            undeployResult = RedisDeployUtils.undeployProxyNode(proxyJson, false, logKey, magicKey, result);
            break;
        case FixDefs.CMPT_REDIS_NODE:
            JsonObject redisJson = DeployUtils.getSpecifiedItem(redisNodeArr, instID);
            undeployResult = RedisDeployUtils.undeployRedisNode(redisJson, true, logKey, magicKey, result);
            break;
        case FixHeader.HEADER_COLLECTD:
            JsonObject collectd = servJson.getJsonObject(FixHeader.HEADER_COLLECTD);
            undeployResult = CollectdDeployUtils.undeployCollectd(collectd, logKey, magicKey, result);
            break;
        default:
            break;
        }
        
        DeployUtils.postDeployLog(undeployResult, servInstID, logKey, "undeploy");
        return undeployResult;
    }

    @Override
    public boolean maintainInstance(String servInstID, String instID, String servType, InstanceOperationEnum op,
            boolean isOperateByHandle, String logKey, String magicKey, ResultBean result) {
        return false;
    }

    @Override
    public boolean updateInstanceForBatch(String servInstID, String instID, String servType, boolean loadDeployFile,
            boolean rmDeployFile, boolean isOperateByHandle, String logKey, String magicKey, ResultBean result) {
        return false;
    }

    @Override
    public boolean checkInstanceStatus(String servInstID, String instID, String servType, String magicKey,
            ResultBean result) {
        return false;
    }

}

File: paas-metasvr-go/pkg/autodeploy/deployer/redis_cluster_deployer.go
package deployer

import (
	"fmt"

	"github.com/maoge/paas-metasvr-go/pkg/consts"
	"github.com/maoge/paas-metasvr-go/pkg/dao/metadao"
	"github.com/maoge/paas-metasvr-go/pkg/deployutils"
	"github.com/maoge/paas-metasvr-go/pkg/deployutils/common"
	RedisDeployUtils "github.com/maoge/paas-metasvr-go/pkg/deployutils/redis"
	"github.com/maoge/paas-metasvr-go/pkg/global"
	"github.com/maoge/paas-metasvr-go/pkg/meta"
	"github.com/maoge/paas-metasvr-go/pkg/result"
	"github.com/maoge/paas-metasvr-go/pkg/utils"
)

type RedisClusterDeployer struct {
}

func (h *RedisClusterDeployer) DeployService(servInstID, deployFlag, logKey, magicKey string, paasResult *result.ResultBean) bool {
	if !deployutils.GetServiceTopo(servInstID, logKey, paasResult) {
		return false
	}

	serv := meta.CMPT_META.GetService(servInstID)
	if deployutils.IsServiceDeployed(logKey, serv, paasResult) {
		return false
	}

	inst := meta.CMPT_META.GetInstance(servInstID)
	cmpt := meta.CMPT_META.GetCmptById(inst.CMPT_ID)

	topoJson := paasResult.RET_INFO.(map[string]interface{})
	paasResult.RET_INFO = ""

	servJson := topoJson[cmpt.CMPT_NAME].(map[string]interface{})

	if deployFlag == consts.DEPLOY_FLAG_PSEUDO {
		if !RedisDeployUtils.RedisClusterServiceFakeDeploy(servJson, servInstID, logKey, magicKey, consts.STR_DEPLOY, paasResult) {
			global.GLOBAL_RES.PubFailLog(logKey, "redis deploy failed ......")
			return false
		}

		info := fmt.Sprintf("service inst_id:%s, deploy sucess ......", servInstID)
		global.GLOBAL_RES.PubSuccessLog(logKey, info)
		return true
	}

	version := serv.VERSION
	nodeContainer := servJson[consts.HEADER_REDIS_NODE_CONTAINER].(map[string]interface{})
	proxyContainer := servJson[consts.HEADER_REDIS_PROXY_CONTAINER].(map[string]interface{})

	redisNodeArr := nodeContainer[consts.HEADER_REDIS_NODE].([]map[string]interface{})
	proxyArr := proxyContainer[consts.HEADER_REDIS_PROXY].([]map[string]interface{})

	node4cluster, nodes4proxy := RedisDeployUtils.GetClusterNodes(&redisNodeArr)

	utils.LOGGER.Info(node4cluster)
	utils.LOGGER.Info(nodes4proxy)

	// 1.deploy redis nodes
	init := false
	size := len(redisNodeArr)
	for idx, redisNode := range redisNodeArr {
		if idx == size-1 {
			init = true
		}

		if !RedisDeployUtils.DeployRedisNode(redisNode, init, false, true, false, node4cluster, version, logKey, magicKey, paasResult) {
			return false
		}
	}

	// 2.deploy proxy
	for _, proxy := range proxyArr {
		if !RedisDeployUtils.DeployProxyNode(proxy, nodes4proxy, logKey, magicKey, paasResult) {
			return false
		}
	}

	// 3.deploy collectd
	collectdRaw := servJson[consts.HEADER_COLLECTD]
	if collectdRaw != nil {
		collectd := collectdRaw.(map[string]interface{})
		if len(collectd) > 0 {
			if !common.DeployCollectd(collectd, servInstID, logKey, magicKey, paasResult) {
				return false
			}
		}
	}

	// 4.update t_meta_service.is_deployed and local cache
	if !metadao.UpdateInstanceDeployFlag(servInstID, consts.STR_TRUE, logKey, magicKey, paasResult) {
		return false
	}
	if !metadao.UpdateServiceDeployFlag(servInstID, consts.STR_TRUE, logKey, magicKey, paasResult) {
		return false
	}

	info := fmt.Sprintf("service inst_id:%s, deploy sucess ......", servInstID)
	global.GLOBAL_RES.PubSuccessLog(logKey, info)

	return true
}

func (h *RedisClusterDeployer) UndeployService(servInstID string, force bool, logKey string, magicKey string,
	paasResult *result.ResultBean) bool {

	if !deployutils.GetServiceTopo(servInstID, logKey, paasResult) {
		return false
	}

	serv := meta.CMPT_META.GetService(servInstID)
	// 未部署直接退出不往下执行
	if deployutils.IsServiceNotDeployed(logKey, serv, paasResult) {
		return false
	}

	inst := meta.CMPT_META.GetInstance(servInstID)
	cmpt := meta.CMPT_META.GetCmptById(inst.CMPT_ID)

	topoJson := paasResult.RET_INFO.(map[string]interface{})
	paasResult.RET_INFO = ""

	servJson := topoJson[cmpt.CMPT_NAME].(map[string]interface{})

	// 卸载伪部署
	if serv.PSEUDO_DEPLOY_FLAG == consts.DEPLOY_FLAG_PSEUDO {
		if !RedisDeployUtils.RedisClusterServiceFakeDeploy(servJson, servInstID, logKey, magicKey, consts.STR_UNDEPLOY, paasResult) {
			global.GLOBAL_RES.PubFailLog(logKey, "redis undeploy failed ......")
			return false
		}

		info := fmt.Sprintf("service inst_id: %s, undeploy sucess ......", servInstID)
		global.GLOBAL_RES.PubSuccessLog(logKey, info)
		return true
	}

	nodeContainer := servJson[consts.HEADER_REDIS_NODE_CONTAINER].(map[string]interface{})
	proxyContainer := servJson[consts.HEADER_REDIS_PROXY_CONTAINER].(map[string]interface{})

	redisNodeArr := nodeContainer[consts.HEADER_REDIS_NODE].([]map[string]interface{})
	proxyArr := proxyContainer[consts.HEADER_REDIS_PROXY].([]map[string]interface{})

	// 1.undeploy proxy
	for _, proxyJson := range proxyArr {
		if !RedisDeployUtils.UndeployProxyNode(proxyJson, force, logKey, magicKey, paasResult) {
			if !force {
				return false
			}
		}
	}

	// 2.undeploy redis nodes
	for _, redisJson := range redisNodeArr {
		if !RedisDeployUtils.UndeployRedisNode(redisJson, false, logKey, magicKey, paasResult) {
			if !force {
				return false
			}
		}
	}
	// 3.卸载collectd服务
	collectdRaw := servJson[consts.HEADER_COLLECTD]
	if collectdRaw != nil {
		collectd := collectdRaw.(map[string]interface{})
		if len(collectd) > 0 {
			if !common.UndeployCollectd(collectd, logKey, magicKey, paasResult) {
				return false
			}
		}
	}

	// 4.update t_meta_service is_deployed flag
	if !metadao.UpdateInstanceDeployFlag(servInstID, consts.STR_FALSE, logKey, magicKey, paasResult) {
		return false
	}
	if !metadao.UpdateServiceDeployFlag(servInstID, consts.STR_FALSE, logKey, magicKey, paasResult) {
		return false
	}

	info := fmt.Sprintf("service inst_id: %s, undeploy sucess ......", servInstID)
	global.GLOBAL_RES.PubSuccessLog(logKey, info)

	return true
}

func (h *RedisClusterDeployer) DeployInstance(servInstID string, instID string, logKey string, magicKey string,
	paasResult *result.ResultBean) bool {

	instCmpt := meta.CMPT_META.GetInstanceCmpt(instID)
	if instCmpt == nil {
		errMsg := fmt.Sprintf("instance %s component not found ......", instID)
		global.GLOBAL_RES.PubFailLog(logKey, errMsg)

		paasResult.RET_CODE = consts.REVOKE_NOK
		paasResult.RET_INFO = errMsg
		return false
	}

	if !deployutils.GetServiceTopo(servInstID, logKey, paasResult) {
		return false
	}

	servInst := meta.CMPT_META.GetInstance(servInstID)
	serv := meta.CMPT_META.GetService(servInstID)
	version := serv.VERSION

	topoJson := paasResult.RET_INFO.(map[string]interface{})
	paasResult.RET_INFO = ""

	servCmpt := meta.CMPT_META.GetCmptById(servInst.CMPT_ID)
	servJson := topoJson[servCmpt.CMPT_NAME].(map[string]interface{})

	nodeContainer := servJson[consts.HEADER_REDIS_NODE_CONTAINER].(map[string]interface{})
	proxyContainer := servJson[consts.HEADER_REDIS_PROXY_CONTAINER].(map[string]interface{})

	redisNodeArr := nodeContainer[consts.HEADER_REDIS_NODE].([]map[string]interface{})
	proxyArr := proxyContainer[consts.HEADER_REDIS_PROXY].([]map[string]interface{})

	node4cluster, nodes4proxy := RedisDeployUtils.GetClusterNodes(&redisNodeArr)

	if instCmpt.CMPT_NAME == consts.CMPT_REDIS_PROXY {
		for _, proxyJson := range proxyArr {
			currId := proxyJson[consts.HEADER_INST_ID]
			if currId != instID {
				continue
			}

			if !RedisDeployUtils.DeployProxyNode(proxyJson, nodes4proxy, logKey, magicKey, paasResult) {
				return false
			}
			// TODO notify clients ......
		}
	} else if instCmpt.CMPT_NAME == consts.CMPT_REDIS_NODE {
		for _, redisJson := range redisNodeArr {
			currId := redisJson[consts.HEADER_INST_ID]
			if currId != instID {
				continue
			}

			if !RedisDeployUtils.DeployRedisNode(redisJson, false, false, true, true, node4cluster, version, logKey, magicKey, paasResult) {
				return false
			}
		}
	} else if instCmpt.CMPT_NAME == consts.CMPT_COLLECTD {
		//卸载collectd服务
		collectdRaw, found := servJson[consts.HEADER_COLLECTD]
		if found {
			collectd := collectdRaw.(map[string]interface{})
			if !common.UndeployCollectd(collectd, logKey, magicKey, paasResult) {
				return false
			}
		}
	}

	return true
}

func (h *RedisClusterDeployer) UndeployInstance(servInstID string, instID string, logKey string, magicKey string,
	paasResult *result.ResultBean) bool {

	instCmpt := meta.CMPT_META.GetInstanceCmpt(instID)
	if instCmpt == nil {
		errMsg := fmt.Sprintf("instance %s component not found ......", instID)
		global.GLOBAL_RES.PubFailLog(logKey, errMsg)

		paasResult.RET_CODE = consts.REVOKE_NOK
		paasResult.RET_INFO = errMsg
		return false
	}

	if !deployutils.GetServiceTopo(servInstID, logKey, paasResult) {
		return false
	}

	servInst := meta.CMPT_META.GetInstance(servInstID)

	topoJson := paasResult.RET_INFO.(map[string]interface{})
	paasResult.RET_INFO = ""

	servCmpt := meta.CMPT_META.GetCmptById(servInst.CMPT_ID)
	servJson := topoJson[servCmpt.CMPT_NAME].(map[string]interface{})

	nodeContainer := servJson[consts.HEADER_REDIS_NODE_CONTAINER].(map[string]interface{})
	proxyContainer := servJson[consts.HEADER_REDIS_PROXY_CONTAINER].(map[string]interface{})

	redisNodeArr := nodeContainer[consts.HEADER_REDIS_NODE].([]map[string]interface{})
	proxyArr := proxyContainer[consts.HEADER_REDIS_PROXY].([]map[string]interface{})

	if instCmpt.CMPT_NAME == consts.CMPT_REDIS_PROXY {
		for _, proxyJson := range proxyArr {
			currId := proxyJson[consts.HEADER_INST_ID]
			if currId != instID {
				continue
			}

			if !RedisDeployUtils.UndeployProxyNode(proxyJson, false, logKey, magicKey, paasResult) {
				return false
			}
			// TODO notify clients ......
		}
	} else if instCmpt.CMPT_NAME == consts.CMPT_REDIS_NODE {
		for _, redisJson := range redisNodeArr {
			currId := redisJson[consts.HEADER_INST_ID]
			if currId != instID {
				continue
			}
			if !RedisDeployUtils.UndeployRedisNode(redisJson, true, logKey, magicKey, paasResult) {
				return false
			}
		}
	} else if instCmpt.CMPT_NAME == consts.CMPT_COLLECTD {
		//卸载collectd服务
		collectdRaw, found := servJson[consts.HEADER_COLLECTD]
		if found {
			collectd := collectdRaw.(map[string]interface{})
			if !common.UndeployCollectd(collectd, logKey, magicKey, paasResult) {
				return false
			}
		}
	}

	return true
}
